tpm2: add support for a trusted SRK

williamcroberts opened the pull request:
Prevent attackers from spoofing the tpmKey portion of the AuthSession by adding a trusted key to the LUKS header metadata. Also, use a persistent object rather than a transient object.

This provides the following benifits:
1. No way to MITM the tpmKey portion of the session, see [1] for details.

2. Strengthens the encrypted sessions, note that the bindKey could be dropped now.

3. Speed, once it's created we just use it.

This is a "first to set" model, in where the first person to set the key in the LUKS header wins. Thus, setup should be done in a known good state. If an SRK, which is a primary key at a special persistent address, is found, it will use whatever is there. If not, it creates an SRK. The SRK follows the convetions used through the tpm2-software organization code on GitHub [2], however, a split has occured between Windows and Linux with respect to SRK templates. The Linux SRK is generated with the unique field size set to 0, in Windows, it properly sets the size to key size in bytes and the unique data to all 0's of that size. Note the proper templates for SRKs is covered in spec [3]. However, the most important thing, is that both SRKs are passwordless, and thus they should be interchangable. If Windows is the first to make the SRK, systemd will gladly accept it and vice-versa.

1. Without the bindKey being utilized, an attacker was able to intercept this and fake a key, thus being able to decrypt and encrypt traffic as needed. Introduction of the bindKey strengthened this, but allows for the attacker to brute force AES128CFB using pin guesses. Introduction of the salt increases the difficulty of this attack as well as DA attacks on the TPM objects itself.

2. https://github.com/tpm2-software

3. https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf

Signed-off-by: William Roberts <william.c.roberts@intel.com>

<!-- devel-freezer = {"comment-id":"1421063204","freezing-tag":"v253-rc2"} -->

williamcroberts commented:
It seems as though UBSAN doesn't like the definition of `TPM2_PERSISTENT_FIRST`. Which expands to:
`((TPM2_HC) (((TPM2_HC) (((TPM2_HT) 0x81) << ((TPM2_HC) 24))) + 0))`

williamcroberts commented:
I wonder if it's because TPM2_HT is a UINT8 which is being leftshifted.... looks like that's it, if we promote the UINT8 to a handle size it corrects this behavior. This is a bug in our code, which we can just work around here.

williamcroberts commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                        TPM2_HR_PERSISTENT_FIRST_GOOD,
+                        TPM2_MAX_CAP_HANDLES,
+                        &more_data,
+                        &cap_data);
```
This is stupid, I should just ask the tpm for a list of handles of Len 1 starting at SRK. If the retuned list is empty, not there, else present... duh.

medhefgo commented on src/test/test-tpm2.c:
```diff
@@ -69,4 +69,95 @@ TEST(tpm2_util_pbkdf2_hmac_sha256) {
+                },
+        };
+
+        for (unsigned i = 0; i < sizeof(templ)/sizeof(templ[0]); i++) {
```
```suggestion
        for (unsigned i = 0; i < ELEMENTSOF(templ); i++) {
```

poettering replied:
also, ELEMENTSOF() returns a `size_t` constant, hence please change typoe of `i` to `size_t` too.

medhefgo commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                                       "Failed to enumerate handles searching for SRK: %s",
+                                       sym_Tss2_RC_Decode(rc));
+
+        for (UINT32 i=0; i < cap_data->data.handles.count; i++) {
```
Wonder where UINT32 is coming from. Please use stdint types.

poettering replied:
i guess cap_data->data.handles.count uses that type...

but yeah, uint32_t or size_t should be used here I guess.

williamcroberts replied:
This goes away in my revised patch where we no longer have to search the list.

poettering commented:
/cc @ddstreet

poettering commented on src/cryptenroll/cryptenroll-tpm2.c:
```diff
@@ -142,7 +142,9 @@ int enroll_tpm2(struct crypt_device *cd,
         _cleanup_(erase_and_freep) char *base64_encoded = NULL;
-        size_t secret_size, blob_size, hash_size, pubkey_size = 0;
+        _cleanup_(freep) void *primary_buf = NULL;
+
```
spurious empty line?

poettering commented on src/cryptsetup/cryptsetup-tokens/cryptsetup-token-systemd-tpm2.c:
```diff
@@ -234,6 +240,7 @@ _public_ void cryptsetup_token_dump(
         crypt_log(cd, "\ttpm2-policy-hash:" CRYPT_DUMP_LINE_SEP "%s\n", policy_hash_str);
         crypt_log(cd, "\ttpm2-pin:         %s\n", true_false(flags & TPM2_FLAGS_USE_PIN));
         crypt_log(cd, "\ttpm2-salt:        %s\n", true_false(salt));
+        crypt_log(cd, "\ttpm2-srk:         %s\n", true_false(primary_buf));
```
hmm, so towards the user the concept is called "srk", and internally we call it "primary"? I think we should stick to one name here.

poettering commented on src/shared/creds-util.c:
```diff
@@ -1046,6 +1048,9 @@ int decrypt_credential_and_warn(
                                 le32toh(t->blob_size),
                                 t->policy_hash_and_blob + le32toh(t->blob_size),
                                 le32toh(t->policy_hash_size),
+                                // TODO BILL WHAT TO PLUMB HERE
```
i guess we need to extend the credential structure to make this work. can be done later. maybe add an explanatory comment here, and add an entry to the TODO file

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
-                ESYS_TR *ret_primary,
-                TPMI_ALG_PUBLIC alg,
-                TPMI_ALG_PUBLIC *ret_alg) {
+const TPM2B_PUBLIC *tpm2_get_primary_template(bool use_srk_model, bool is_ecc) {
```
maybe make this a flags param?

```c
typedef enum Tpm2SRKTemplateFlags = {
        TPM2_SRK_TEMPLATE_NEW_STYLE = 1 << 0,
        TPM2_SRK_TEMPLATE_ECC       = 1 << 1,
        _TPM2_SRC_TEMPLATE_MAX      = TPM2_SRK_TEMPLATE_NEW_STYLE|TPM2_SRK_TEMPLATE_ECC,
} Tpm2SRKTemplateFlags;
```

And then:

```c
const TPM2B_PUBLIC *tpm2_get_primary_template(Tpm2SRKTemplateFlags flags) {
        static const TPM2B_PUBLIC templ[_TPM2_SRC_TEMPLATE_MAX] = {
                …
        };
        assert(flags >= 0);
        assert(flags < _TPM2_SRC_TEMPLATE_MAX);
        return templates[flags];  
 }
```

this of course is only efficient if the template space is dense, i.e. for each flags combo we have a distinct template. But so far we have...

using a flags concept makes calling this function a lot more readable as flags names are not as opaque as just "true" and "false"

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary);
```
as per coding style: if we break function parameter lists, please indent to double the usual indentation, i.e. 16 spaces.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary);
+                if (rc)
```
if (rc != TSS2_RC_SUCCESS)

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+            ESYS_TR_NONE,
+            &out_public,
+            &out_name,
+            &out_qname);
```
16 space indenting, please

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+            &out_public,
+            &out_name,
+            &out_qname);
+        if (rc)
```
as above

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -311,55 +321,222 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                        log_debug("Discovered SRK using that");
+                        return 0;
+                }
+                log_debug("Did not find SRK, generating...");
```
hmm, how does tpm2 access locking actually work? i.e. if multiple instances of this code run in parallel, will they just work? or will possibly both detect that an SRK is missing, and both create it in parallel, and once wins and the othe rfails?

williamcroberts replied:
Is this holding /dev/tpmrm0? if so it's per-command transaction, if it's holding /dev/tpm0, then it's exclusive.

williamcroberts replied:
But either way if this code can run in parallel it will need a lock here, whats the mechanism for that in systemd?

poettering replied:
it's /dev/tpmrm0.

So I did get this right: it's one step to check if the SRK is existant, and another one to then generate it, and we need to manually lock around it?

note that cryptsetup instances for different volumes run entirely in parallel. so if there's stuff they should not do in parallel, we need a system-wide mutex for that. Typically we use "unposix" or BSD locks for that (since they are per-fd, auto-cleaning and robust) on some suitable file. This could be the /dev/tpmrm0 fd itself actually, but not sure if the tss2 libs allow us access to the fd it internally uses. Also, it might be too coarse to just say "lock the whole device for everyone", and we'd much rather have a more finer grained lock for just "lock the init srk operation only".

assuming we want the latter, you could use the existing make_lock_file() helper, with a path such as "/run/systemd/tpm2-srk-init" or so.

williamcroberts replied:
> it's /dev/tpmrm0.
> 
> So I did get this right: it's one step to check if the SRK is existant, and another one to then generate it, and we need to manually lock around it?
> 

Yes, 100% correct.

> note that cryptsetup instances for different volumes run entirely in parallel. so if there's stuff they should not do in parallel, we need a system-wide mutex for that. Typically we use "unposix" or BSD locks for that (since they are per-fd, auto-cleaning and robust) on some suitable file. This could be the /dev/tpmrm0 fd itself actually, but not sure if the tss2 libs allow us access to the fd it internally uses. 

GetPollHandles

> Also, it might be too coarse to just say "lock the whole device for everyone", and we'd much rather have a more finer grained lock for just "lock the init srk operation only".

Yes too coarse IMO.

> assuming we want the latter, you could use the existing make_lock_file() helper, with a path such as "/run/systemd/tpm2-srk-init" or so.

Perfect, thanks.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1590,6 +1780,20 @@ int tpm2_seal(const char *device,
+         * the key we use later is the key we expect to establish the session with.
+         */
+        if (ret_primary_buf) {
+                log_debug("Serializing SRK");
```
this log message is a bit misleading, no? after all you cannot extract the SRK, and cannot serialize it. You only serialize the reference to it here. hence maybe say that

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1683,9 +1904,23 @@ int tpm2_unseal(const char *device,
+                                               "Failed to unmarshal primary key: %s", sym_Tss2_RC_Decode(rc));
+        /* old callers without an SRK still need to create a key */
+        } else {
+                r = tpm2_make_primary(c.esys_context, &primary, primary_alg, false, NULL);
```
maybe tpm2_make_primary() should now be renamed to `tpm2_make_transient_primary()` to make clear what this is about

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1788,7 +2023,9 @@ int tpm2_unseal(const char *device,
 finish:
-        primary = tpm2_flush_context_verbose(c.esys_context, primary);
+        if (!primary_buf)
+                primary = tpm2_flush_context_verbose(c.esys_context, primary);
```
why the conditoinalization?

williamcroberts replied:
> why the conditoinalization?

You don't flush a persistent handle, it causes an error.

poettering replied:
ah, makes sense. can you add a comment about that?

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -2288,6 +2531,15 @@ int tpm2_parse_luks2_json(
         } else if (pubkey_pcr_mask != 0)
                 return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Public key PCR mask set, but not public key included in JSON data, refusing.");
 
+
```
spurious double empty line

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -2288,6 +2531,15 @@ int tpm2_parse_luks2_json(
+                r = json_variant_unbase64(w, &primary_buf, &primary_buf_size);
+                if (r < 0)
+                        return log_debug_errno(r, "Invalid base64 data in 'tpm2-blob' field.");
+
```
this empty line seems unnecessary, too

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -2288,6 +2531,15 @@ int tpm2_parse_luks2_json(
+        if (w) {
+                r = json_variant_unbase64(w, &primary_buf, &primary_buf_size);
+                if (r < 0)
+                        return log_debug_errno(r, "Invalid base64 data in 'tpm2-blob' field.");
```
comment needs updating

poettering commented:
hmm, so what's the argument for persisting the primary key as SRK? that's merely about speed? or anything else?

poettering commented:
so this only pins the srk in the json luks2 header. but we could also pin the transient primary key if we have it, couldn't we? it's also derived from the seed key, hence should remain fixed for the same tpm2 chip, no?

poettering commented:
i wonder if we should also do the SRK initialization independently of cryptsetup in some new tiny binary `systemd-provision-tpm2` or so. For two reasons: this is slow, right? so we could do it super early, long before the disks have popped up. And thus when they do pop up we don't have to wait for that step anymore. 

But more interestingly: it would be independent from disk encryption, i.e. would be usable for the creds stuff or any other tpm2 user the same way.

such a mini service would just trivially call the code you just added.

does that make sense? (just asking generally, not suggesting you add this to this PR)

poettering commented:
(somewhat related https://github.com/systemd/systemd/issues/20668)

williamcroberts commented:
> hmm, so what's the argument for persisting the primary key as SRK? that's merely about speed? or anything else?

1. it follows the spec https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
  - This is how it should be done. This will allow all the projects to interoperate.
2. We get a constant known trusted key in the TPM
   - Yes make primary returns a consistent object, but why do it every call.
3. speed

williamcroberts commented:
> i wonder if we should also do the SRK initialization independently of cryptsetup in some new tiny binary `systemd-provision-tpm2` or so. For two reasons: this is slow, right?

Creating keys is slow, (ecc is faster) but the evictcontrol is usually fast.

 > so we could do it super early, long before the disks have popped up. And thus when they do pop up we don't have to wait for that step anymore.

Yes you could.

> 
> But more interestingly: it would be independent from disk encryption, i.e. would be usable for the creds stuff or any other tpm2 user the same way.
> 
> such a mini service would just trivially call the code you just added.
> 
> does that make sense?
Yes, and in theory that's possible now. I could create the SRK with tpm2-tools or Windows, etc, upgrade my systemd, cryptenroll a new device and have it pick up that SRK.

>  (just asking generally, not suggesting you add this to this PR)
Thank goodness, I don't know if I have the systemd chops to turn that around in a reasonable timeframe.

williamcroberts commented:
> so this only pins the srk in the json luks2 header. but we could also pin the transient primary key if we have it, couldn't we? it's also derived from the seed key, hence should remain fixed for the same tpm2 chip, no?

Yes you could pin it as well.

williamcroberts commented:
> > so this only pins the srk in the json luks2 header. but we could also pin the transient primary key if we have it, couldn't we? it's also derived from the seed key, hence should remain fixed for the same tpm2 chip, no?
> 
> Yes you could pin it as well.

The code is not as simple, because an ESYS_TR is combination of a name + handle, so you'd have to just store the name and manually do a name check where it matters since it's not a stable handle. Right now, but just serializing to the SRK and deserialzing the ESYS_TR, ESAPI automatically can perform name checking without client code being responsible for it.

williamcroberts commented:
> > hmm, so what's the argument for persisting the primary key as SRK? that's merely about speed? or anything else?
> 
> 1. it follows the spec https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
> 
> * This is how it should be done. This will allow all the projects to interoperate.
> 
> 2. We get a constant known trusted key in the TPM
>    
>    * Yes make primary returns a consistent object, but why do it every call.
> 3. speed

@poettering I forgot the most important reason, only someone with owner authorization can create a key in the owner hierarchy, all this code is currently broken, we need to prompt for owner password if it's set (getcap call enables this). 

With this patch, their's at least a workaround, create the SRK first with tpm2-tools and then perform cryptenroll.

williamcroberts commented:
I've done a bad job explaining this all in one cohesive thought. The idea is that the TPM Owner might not be the computer owner/admin.  I.E. the person performing cryptenroll might not have owner auth. In the current model, the owner auth has to be empty, which means that if I as a regular user want to do something, I would need to clear owner auth and re-establish it after the task is done. This is the WHY the SRK exists. The SRK exists to be a scratch space for regular users, which is why the password is emptyauth. This patch at least gets it to a point where the TPM Owner can create the SRK and walk away and users can cryptenroll to their hearts content.

poettering commented:
that explanation for the SRK would be good in the code btw, where we generate it. TPM2 stuff is so opaque, would be more than excellent to have that right in the code where epople can see it. More or less just copy what you wrote here into a code comment maybe.

poettering commented:
Btw, might make sense to reference #22637 from your commit msg, as you implement a part of what's requested there.

> @poettering I forgot the most important reason, only someone with owner authorization can create a key in the owner hierarchy, all this code is currently broken, we need to prompt for owner password if it's set (getcap call can get this information). So any current OWNER password set TPMs, this all breaks IIUC.

I figure that's issue #22129?

poettering commented:
And I guess you could add "Fixes: #20668", given with this we'll have NODA for new enrollments.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -307,55 +317,206 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+
+        static const TPM2B_PUBLIC templ[] = {
+                /* index 0 RSA old */
+                {
```
i'd make this more explicit, i.e. add here:

```c
static const TPM2B_PUBLIC templ[_TPM2_SRC_TEMPLATE_MAX] = {
        [0] = { 
                …
        },
        [TPM2_SRK_TEMPLATE_ECC] = {
                …
        },
        [TPM2_SRK_TEMPLATE_NEW_STYLE] = { 
                …
        },
        [TPM2_SRK_TEMPLATE_NEW_STYLE|TPM2_SRK_TEMPLATE_ECC] = {
                …
        }
};

williamcroberts replied:
This is 3, when the len is 4, so either I need to add one to `_TPM2_SRC_TEMPLATE_MAX` or 1 to the array definition, do you have a preference?

williamcroberts replied:
For now I just did a `+ 1` at array definitions and performed less than or equal tests on incoming flags to the constant.

poettering replied:
yeah, that works.

poettering commented on src/test/test-tpm2.c:
```diff
@@ -69,4 +69,94 @@ TEST(tpm2_util_pbkdf2_hmac_sha256) {
+                                },
+                        },
+                },
+        };
```
hmm, these structures are the same on all archs, right?

so instead of defining the whole thing twice, maybe just put a list of sha hashes here and validate them? we have sha256 code in place anyway, so that would be easy?

i think that would be less ugly then to literally write down the whole structs twice

williamcroberts replied:
I think this would be the first time I disagree, but if it's a show stopper I'll do it. So instead of visually being able to see the comparison to whats expected, and then understand the changes implications, it's to magic hashes that need to be explained.

williamcroberts replied:
I added this commit on top of the base commit, we can make a call to take it as a separate change or squash it down, just let me know.

poettering replied:
hmm, so these structs, are not actually the same everywhere, are they? i.e. this is not what is sent over the wire 1:1, right? because if it was, then we'd have to do a lot of htole()/htobe() stuff here... But that means the hashes are not stable anyway, as they would be different on LE and BE machines.

so i guess scratch my hash-based checking idea anyway...

williamcroberts replied:
> hmm, so these structs, are not actually the same everywhere, are they? i.e. this is not what is sent over the wire 1:1, right? because if it was, then we'd have to do a lot of htole()/htobe() stuff here... But that means the hashes are not stable anyway, as they would be different on LE and BE machines.
> 
> so i guess scratch my hash-based checking idea anyway...

Correct, scalars are host to big endian converted.

poettering commented:
looks pretty good

williamcroberts commented:
> And I guess you could add "Fixes: #20668", given with this we'll have NODA for new enrollments.

Looks like it fixes #20668 and #22637

williamcroberts commented:
> Btw, might make sense to reference #22637 from your commit msg, as you implement a part of what's requested there.
> 
> > @poettering I forgot the most important reason, only someone with owner authorization can create a key in the owner hierarchy, all this code is currently broken, we need to prompt for owner password if it's set (getcap call can get this information). So any current OWNER password set TPMs, this all breaks IIUC.
> 
> I figure that's issue #22129?

Yep, that's the one. If we can migrate all users to the SRK that problem goes away once SRKs just exist everywhere.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -307,55 +317,213 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+                ESYS_CONTEXT *c,
+                ESYS_TR *ret_primary,
+                TPMI_ALG_PUBLIC alg,
+                TPMI_ALG_PUBLIC *ret_alg) {
```
in all recent code we put the input params first, the return params last. Hence this would suggest and order of c, lal ret_primary, ret_alg...

williamcroberts replied:
Damn, I thought I got all these!

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -307,55 +317,213 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+
+        /* Did Not find SRK, indicate this by returning 0 and ESYS_TR_NONE */
+        if (cap_data->data.handles.count == 0) {
+                *ret_primary = ESYS_TR_NONE;
```
so we have this rule in systemd that all output params of a function must be initialized on success, and must be left untouched on error. this here is a success path, hence should initialize `*ret_alg` to something (to 0 or so?)

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -307,55 +317,213 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
-        log_debug("Creating primary key on TPM.");
+        /* Find existing SRK and use it if present */
+        if (use_srk_model) {
+                rc = tpm2_get_srk(c, ret_primary, alg, ret_alg);
```
so as mentioned we have this rule that we want that for each function all return params are fully initiallized on success or not touched at all on failure.

now, here you pass ret_primary/ret_alg directly into tpm2_get_srk(), which will hence be modified on success. which is fine. but there are error paths further down here which might still fail, and if they do then you touched return params even on failure...

hence please use a local variable first, and only if the SRK already exists copy it from there into ret_primary/ret_alg.

(you already have `primary` and `alg` as local vars, so just use those)

williamcroberts replied:
I also added asserts here and elsewhere now on things that callers cannot optionally pass for pointers.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -307,55 +317,213 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+
+        static const TPM2B_SENSITIVE_CREATE primary_sensitive = {};
+        const TPM2B_PUBLIC *primary_template = NULL;
+
```
spurious empty line?

williamcroberts replied:
I have must have misread the diff, I actually added that line back in :-p... removing.

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1586,6 +1767,20 @@ int tpm2_seal(const char *device,
+        if (ret_srk_buf) {
+                log_debug("Serializing SRK ESYS_TR reference");
+                rc = sym_Esys_TR_Serialize(c.esys_context, primary, &srk_buf, &srk_buf_size);
+                if (rc) {
```
rc != TSS2_RC_SUCCESS

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1598,10 +1793,25 @@ int tpm2_seal(const char *device,
+
+                *ret_srk_buf = tmp;
+                *ret_srk_buf_size = srk_buf_size;
+        }
```
like int the other cases, if the memdup() here fails you already touched ret_secret and all the other stuff. hence please move this block just a bit up, *before* the `*ret_secret = TAKE_PTR(…` line, and things are in order again

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1679,9 +1891,23 @@ int tpm2_unseal(const char *device,
+                                               "Failed to unmarshal primary key: %s", sym_Tss2_RC_Decode(rc));
+        /* old callers without an SRK still need to create a key */
+        } else {
+                r = tpm2_make_transient_primary(c.esys_context, &primary, primary_alg, false, NULL);
```
i recent code we like if opaque arguments in function calls, such as "true", "false" or NULL, or 0 are prefixed with the name of the argument, since from the values you pass you otherwise have no idea what this might be.

hence maybe:

```c
r = tpm2_make_transient_primary(
                c.esys_context, 
                &primary, 
                primary_alg, 
                /* use_srk_model= */ false, 
                /* ret_arg= */ NULL);

poettering commented on src/test/test-tpm2.c:
```diff
@@ -69,4 +69,94 @@ TEST(tpm2_util_pbkdf2_hmac_sha256) {
+                        },
+                },
+        };
+
```
maybe also add:

```c
assert_cc(ELEMENTSOF(templ) == _TPM2_SRK_TEMPLATE_MAX+1);
```

williamcroberts commented on src/shared/tpm2-util.c:
```diff
@@ -2133,6 +2363,7 @@ int tpm2_make_luks2_json(
                         return r;
         }
 
+
```
spurious, removing.

williamcroberts commented on src/shared/tpm2-util.c:
```diff
@@ -307,55 +318,227 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
-        log_debug("Creating primary key on TPM.");
+        assert(primary);
+
+        int r = make_lock_file("/run/systemd/tpm2-srk-init", LOCK_EX, &srk_lock);
```
I guess this lock only needs to be on the srk path.... changing.

williamcroberts commented:
@poettering I think I got everything, FYI if you really want the tests on sha256sum I could use libmu to marshal the TPM2B_PUBLIC structures to the buffer as sent to the TPM (scalars converted to big endian), so the hashes would be stable. I think that's overkill however :-p

ddstreet commented:
Well, my PR (#26027) is gonna massively collide with this :(

I think as @poettering [commented](https://github.com/systemd/systemd/pull/26185#issuecomment-1403411372) that  this shouldn't ever be done implicitly; persisting the SRK should be an explicit choice by the system admin, either my doing it manually with tpm2_evictcontrol or similar, or we could create a simple program to do it (easier than the manual steps using tpm2_* tools). That way, you don't have to worry about any locking, but also implicitly creating the SRK might not be what the system owner actually wants.

Could you delay this until more of my PR is merged? I think it's going to be a lot easier and more flexible once some of my changes are in...otherwise I think I would have to rebase on this and probably redo a lot of it...

williamcroberts commented:
> Well, my PR (#26027) is gonna massively collide with this :(
> 
> I think as @poettering [commented](https://github.com/systemd/systemd/pull/26185#issuecomment-1403411372) that this shouldn't ever be done implicitly; persisting the SRK should be an explicit choice by the system admin, either my doing it manually with tpm2_evictcontrol or similar, or we could create a simple program to do it (easier than the manual steps using tpm2_* tools). 

I think the comment was around decoupling the steps,
"does that make sense? (just asking generally, not suggesting you add this to this PR)".

Almost everything, including Windows, if it needs the SRK just creates it. But a TPM admin could create the SRK, or if they don't want one, never provide the owner auth so no one could enroll to a tpm. This allows a workaround to the current model always requiring owner auth.

> That way, you don't have to worry about any locking, but also implicitly creating the SRK might not be what the system owner actually wants.
> 

Theirs two owners here:
1. Tpm owner, who still needs to provide owner auth to create a primary key
2. The system admin, can control who has tpm access.

> Could you delay this until more of my PR is merged? 

I mean this is ready now ;-p

I think it's going to be a lot easier and more flexible once some of my changes are in...otherwise I think I would have to rebase on this and probably redo a lot of it...

You have some issues in your code IIUC on createaloaded, but I didn't look through it in detail.

ElvishJerricco commented:
Can someone please explain / add documentation about what this means to end users? Do I have to worry about TPM ownership now?

williamcroberts commented:
> Can someone please explain / add documentation about what this means to end users? Do I have to worry about TPM ownership now?

It doesn't change anything with respect to who holds the owner hierarchy auth. In both cases a tpm2_clear command rolls the owner hierarchy seed and thus nukes any owner hierarchy keys. This commit means that your session encryption used to protect the disc encryption key is verified. You can think of this as verifying the connection over https. Before there was no verification, I added a bind key which helps, but this helps fortify those protections.

If there's somewhere to drop some documentation on this I can add it I just don't know where.

ddstreet commented:
> > Well, my PR (#26027) is gonna massively collide with this :(
> > I think as @poettering [commented](https://github.com/systemd/systemd/pull/26185#issuecomment-1403411372) that this shouldn't ever be done implicitly; persisting the SRK should be an explicit choice by the system admin, either my doing it manually with tpm2_evictcontrol or similar, or we could create a simple program to do it (easier than the manual steps using tpm2_* tools).
> 
> I think the comment was around decoupling the steps, "does that make sense? (just asking generally, not suggesting you add this to this PR)".
> 
> Almost everything, including Windows, if it needs the SRK just creates it. But a TPM admin could create the SRK, or if they don't want one, never provide the owner auth so no one could enroll to a tpm. This allows a workaround to the current model always requiring owner auth.

anyone with owner auth (or if owner auth is empty, which is commonly default) can create an SRK, that's what systemd currently does. the issue is persisting the SRK at its reserved handle.

I completely agree that systemd should *use* the reserved SRK handle, if it's populated, what I don't agree with is systemd implicitly persisting something at the reserved handle. That should be an *explicit* action by someone.

> 
> > That way, you don't have to worry about any locking, but also implicitly creating the SRK might not be what the system owner actually wants.
> 
> Theirs two owners here:
> 
> 1. Tpm owner, who still needs to provide owner auth to create a primary key
> 2. The system admin, can control who has tpm access.

yes, obviously, though in many cases the owner hierarchy has the empty auth so anyone can use it to create a (storage) primary key.

but *we* (i.e. systemd devs) are not either of those owners (besides on our own machines) and I'm not sure that we should be deciding to implicitly persist an SRK (at the reserved handle).

> 
> > Could you delay this until more of my PR is merged?
> 
> I mean this is ready now ;-p
> 
> I think it's going to be a lot easier and more flexible once some of my changes are in...otherwise I think I would have to rebase on this and probably redo a lot of it...
> 
> You have some issues in your code IIUC on createaloaded, but I didn't look through it in detail.

not sure what you're referring to there

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -307,55 +318,232 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
+}
+
+
+static int tpm2_make_transient_primary(
```
it seems wrong to name this 'make transient primary' when it's actually persisting an SRK...

williamcroberts replied:
> it seems wrong to name this 'make transient primary' when it's actually persisting an SRK...

That was @poettering 's suggestion so I didn't fight it.

williamcroberts replied:
I changed it back

ddstreet commented:
> > anyone with owner auth (or if owner auth is empty, which is commonly default) can create an SRK, that's what systemd currently does. the issue is persisting the SRK at its reserved handle.
> 
> While the owner auth is typically not set for a new TPM, it should must be one of the first things changed. An empty owner auth means anyone can just nuke your keys, NV indices, etc. Depending on how NV indices are defined, it could allow owner auth to read/write them. You also want to set up passwords for the other hierarchies.

an empty owner auth isn't necessarily bad.  even with an empty owner auth, it's still impossible to improperly unseal a secret. it is possible to overwrite or read handles, but you can't read the private key of a handle. the tpm will still perform securely re: unsealing even with an empty owner auth.

> 
> > I completely agree that systemd should use the reserved SRK handle, if it's populated, what I don't agree with is systemd implicitly persisting something at the reserved handle. That should be an explicit action by someone.
> 
> It is explicit, it's through the act of cryptenroll. Again this is how Windows does it, which is generally considered a sane way to do it, it should just work.

i think we can agree to disagree on this one :)

> 
> > yes, obviously, though in many cases the owner hierarchy has the empty auth so anyone can use it to create a (storage) primary key.
> 
> Again that shouldn't be true. If the owner auth is empty, you have a worthless TPM. 

this is entirely untrue.

> Hierarchies can also be controlled by policies as well. I don't expect support for that to trickle into convenience tools like cryptenroll. For those users they should create the SRK externally.
> 
> > but we (i.e. systemd devs) are not either of those owners (besides on our own machines) and I'm not sure that we should be deciding to implicitly persist an SRK (at the reserved handle).
> 
> That's a moot argument. The PC Provisioning Guidance is there to tell software what and how they should do it. Systemd isn't making any decision, it's just now following what the guidance documents and other software does. Now we're just making it play nicely with the TPM and how TPMs get configured and deployed in the field without the user needing to be a TPM expert. Without the owner auth, you're not doing anything anyways.
> 
> The current state is that systemd assumes that owner auth is empty and that it can create a primary key every time it needs to seal or unseal something, which is not the case. It's fundamentally broken in how TPMs get deployed (with hierarchy auths set).

that isn't relevant to my opinion though - as i said i agree we should be *using* the SRK, if it exists. what i don't agree with is (silently) *creating* it if it doesn't exist. And if there is an owner auth set, then we can't create an SRK anyway, without providing the auth, which doesn't make sense (IMHO) to add to systemd-cryptenroll.

> 
> We want new users off of this broken transient key model ASAP, so rather than failing if their is a missing SRK, just create it if we can. We could even prompt for owner auth if it's set.

I think calling the transient key model 'broken' is stretching things...

> 
> This patch makes it so that tpm2 integration with systemd is available to a wider audience.

i certainly don't agree with all your assertions, but changing to use an SRK is good and i can just rebase on top of this if it gets merged. i don't want to hold it up.

williamcroberts commented:
> > > yes, obviously, though in many cases the owner hierarchy has the empty auth so anyone can use it to create a (storage) primary key.
> > 
> > 
> > Again that shouldn't be true. If the owner auth is empty, you have a worthless TPM.
> 
> this is entirely untrue.

Anything with TPM2 access can call clear and nuke your keys. I don't want a confused deputy attack on something that has permission to access the tpm2 device to cause a clear, or on a shared system, have someone clear the tpm intentionally or accidentally. Odds are if your owner hierarchy is not set, neither is the endorsement or lockout, which should also be set. In TPM 1.2 days, this was the "take ownership".   

The TPM is a globaly shared hardware device, I don't want someone playing around with tpm2-tools to nuke my keys.

> 
> > We want new users off of this broken transient key model ASAP, so rather than failing if their is a missing SRK, just create it if we can. We could even prompt for owner auth if it's set.
> 
> I think calling the transient key model 'broken' is stretching things...
> 

In a system with a configured owner auth, ie 99.9% of deployed commercial configurations, it wouldn't work. I never seen
an empty owner auth in a real world deployed physical TPM.

> > This patch makes it so that tpm2 integration with systemd is available to a wider audience.
> 
> i certainly don't agree with all your assertions, but changing to use an SRK is good and i can just rebase on top of this if it gets merged. i don't want to hold it up.

What exactly are you proposing, if an SRK doesn't exist, what should cryptenroll do? We need to move users away from a model requiring owner auth.

williamcroberts commented:
> an empty owner auth isn't necessarily bad. even with an empty owner auth, it's still impossible to improperly unseal a secret. it is possible to overwrite or read handles, but you can't read the private key of a handle. the tpm will still perform securely re: unsealing even with an empty owner auth.

As long as that secret isn't in an NV index protected by the owner auth. You're only considering the sealed key. The owner auth controls a lot of things within the TPM:
1. Owner Hierarchy Non Volatile Memory
2. Audit Status
3. Creating Primary Keys
4. Persisting Keys
5. Setting Owner Auth / Policy
6. Changing the Clock
7. Setting Write Lock
8. Changing Owner Primary Seed (tpm2_clear)

Best practice and provisioning guidance IS to set it

williamcroberts commented:
@poettering FYI rebased on current main and fixed conflicts

williamcroberts commented:
So Centos 8 is failing because it is on  `tpm2-tss  2.3.2` which predates `Esys_TR_GetTPMHandle`. That function was added in 2.4.0. I use that to figure out to flush or not. I could add a field to the Tpm2Handle structure noflush and set that. I was hoping to not put that burden on the programmer. Centos 8 is EOL IIUC, do we still need to support this with new systemd?

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -86,7 +93,13 @@ int dlopen_tpm2(void) {
                         DLSYM_ARG(Esys_Unseal),
-                        DLSYM_ARG(Esys_VerifySignature));
+                        DLSYM_ARG(Esys_VerifySignature),
+                        DLSYM_ARG(Esys_TR_FromTPMPublic),
```
can you please keep this sorted?

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -51,6 +52,12 @@ TSS2_RC (*sym_Esys_TR_GetName)(ESYS_CONTEXT *esysContext, ESYS_TR handle, TPM2B_
 TSS2_RC (*sym_Esys_Unseal)(ESYS_CONTEXT *esysContext, ESYS_TR itemHandle, ESYS_TR shandle1, ESYS_TR shandle2, ESYS_TR shandle3, TPM2B_SENSITIVE_DATA **outData) = NULL;
 TSS2_RC (*sym_Esys_VerifySignature)(ESYS_CONTEXT *esysContext, ESYS_TR keyHandle, ESYS_TR shandle1, ESYS_TR shandle2, ESYS_TR shandle3, const TPM2B_DIGEST *digest, const TPMT_SIGNATURE *signature, TPMT_TK_VERIFIED **validation);
+TSS2_RC (*sym_Esys_TR_FromTPMPublic)(ESYS_CONTEXT *esysContext, TPM2_HANDLE tpm_handle, ESYS_TR optionalSession1, ESYS_TR optionalSession2, ESYS_TR optionalSession3, ESYS_TR *object);
+TSS2_RC (*sym_Esys_ReadPublic)(ESYS_CONTEXT *esysContext, ESYS_TR objectHandle, ESYS_TR shandle1, ESYS_TR shandle2, ESYS_TR shandle3, TPM2B_PUBLIC **outPublic, TPM2B_NAME **name, TPM2B_NAME **qualifiedName);
```
can you please keep these sorted (alphabetically)?

williamcroberts replied:
its no longer completely sorted, I can do best effort. Same for above

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -229,11 +242,26 @@ int tpm2_context_new(const char *device, Tpm2Context **ret_context) {
                 return;
 
-        TSS2_RC rc = sym_Esys_FlushContext(esys_context, esys_handle);
+        /* TPM handles are in the ESYS_TR cache, so this is fast and
```
let's keep this logic out of tpm2_handle_flush, this function should be left to *only* flush the handle. checking if it really should be flushed or not can go into tpm2_handle_free, or possibly even higher up

williamcroberts replied:
> this function should be left to only flush the handle

This function already checks if the handle is `ESYS_TR_NONE` and skips flushing it, so checking if it's persistent or not fit the bill. But since that function has to go away and I have to add something to Tpm2Handle to decide if to flush or not, it has to move up a level.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+}
+
+/*
+ * Why and what is an SRK?
```
this is confusing and doesn't clarify at all why the SRK is at that specific reserved handle. it would be better to reduce this paragraph down to 1-2 simpler sentences and provide a link to the TCG guidelines doc which provides much more detailed documentation on the why and how.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                                        "SRK expected alg 0x%x does not match found alg 0x%x",
+                                        alg, out_public->publicArea.type);
+
+        ret_primary->esys_handle = primary;
```
no, you need to TAKE_PTR here

williamcroberts replied:
No primary is an ESYS_TR and is being assigned to the already allocated Tpm2Handle specified from the caller. The calling routine creates a Tpm2Handle which is used either for a transient primary handle or the persistent SRK depending on what the higher level caller supports. This keeps backwards compat for some of the old mechanisms calling into this stuff as mentioned in the TODO IIUC. New, fully supported callers into this end up in the SRK model.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+
+                if (ret_alg)
+                        *ret_alg = 0;
+                return 0;
```
this shouldnt' return 0 on failure to find the SRK

williamcroberts replied:
Ill change this to < 0 to be error, 0 to no SRK, 1 to mean found SRK

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                Tpm2Context *c,
+                TPMI_ALG_PUBLIC alg,
+                TPMI_ALG_PUBLIC *ret_alg,
+                Tpm2Handle *ret_primary) {
```
this should be **ret_primary

williamcroberts replied:
No as it reuses the already created empty Tpm2Handle structure in the calling function

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                Tpm2Handle *ret_primary) {
+
+        TPMI_YES_NO more_data;
+        ESYS_TR primary = ESYS_TR_NONE;
```
you need to use _cleanup_ here, not use ESYS_TR directly; otherwise you might leak this on error condition after the FromTPMPublic call

williamcroberts replied:
No, theirs nothing to leak. ESYS_TR is uint32_t under the hood, the GetCap. This function takes a raw hadle returned by GetCap, converts it to an ESYS_TR and deserializes it, nothing to leak memory wise and nothing to flush either.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+
+        /* Make sure the alg is what we were expecting */
+        _cleanup_(Esys_Freep) TPM2B_PUBLIC *out_public = NULL;
+        _cleanup_(Esys_Freep) TPM2B_NAME *out_name = NULL, *out_qname = NULL;
```
why are you reading the name/qname? it's not being used at all

williamcroberts replied:
Oh I guess ESAPI allows null...

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+static int tpm2_get_srk(
+                Tpm2Context *c,
+                TPMI_ALG_PUBLIC alg,
+                TPMI_ALG_PUBLIC *ret_alg,
```
i don't think it makes sense to both provide the alg and return the alg. the check if the alg matches the srk can easily be done by the caller (if it really should be done).

williamcroberts replied:
It matches the existing semantics of the code its tying into, if alg is 0, then that means the function is free to choose, if the caller chooses then it has to match.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -391,15 +566,45 @@ static int tpm2_make_primary(
+
+        /* Find existing SRK and use it if present */
+        if (use_srk_model) {
+                rc = tpm2_get_srk(c, alg, &alg, primary);
```
this is not ideal to me; either the user wants to use the SRK, or they don't. If they do, then it's not up to them what asym alg the SRK is using. So I'm not sure it really makes sense to fail if the user asks for RSA but the SRK is ECC.

williamcroberts replied:
Sure, ill issue a warning (perhaps notice or info level is better?) and move on. It doesn't matter in the end.

williamcroberts replied:
Updated in the newest as a warning and moves on.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +358,202 @@ static int tpm2_credit_random(Tpm2Context *c) {
+        _cleanup_(release_lock_file) LockFile srk_lock = LOCK_FILE_INIT;
         TSS2_RC rc;
         usec_t ts;
-        int r;
```
is there a whitespace error here?

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -391,15 +566,45 @@ static int tpm2_make_primary(
+
+        /* Find existing SRK and use it if present */
+        if (use_srk_model) {
+                rc = tpm2_get_srk(c, alg, &alg, primary);
```
also tpm2_get_srk returns int, not TSS2_RC; wrong return var

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1672,18 +1918,37 @@ int tpm2_unseal(const char *device,
+                                &primary->esys_handle);
+                if (rc != TSS2_RC_SUCCESS)
+                        return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                               "Failed to unmarshal primary key: %s", sym_Tss2_RC_Decode(rc));
```
this is deserializing, not unmarshalling

DaanDeMeyer commented:
> So Centos 8 is failing because it is on tpm2-tss  2.3.2 which predates Esys_TR_GetTPMHandle. That function was added in 2.4.0. I use that to figure out to flush or not. I could add a field to the Tpm2Handle structure noflush and set that. I was hoping to not put that burden on the programmer. Centos 8 is EOL IIUC, do we still need to support this with new systemd?

CentOS Stream 8 is not EOL so we'd still like to support it

williamcroberts commented:
> > So Centos 8 is failing because it is on tpm2-tss  2.3.2 which predates Esys_TR_GetTPMHandle. That function was added in 2.4.0. I use that to figure out to flush or not. I could add a field to the Tpm2Handle structure noflush and set that. I was hoping to not put that burden on the programmer. Centos 8 is EOL IIUC, do we still need to support this with new systemd?
> 
> CentOS Stream 8 is not EOL so we'd still like to support it

Really? It says Dec 31 2021 here, is that wrong?
  - https://www.centos.org/centos-linux-eol/

mrc0mmand commented:
> > > So Centos 8 is failing because it is on tpm2-tss  2.3.2 which predates Esys_TR_GetTPMHandle. That function was added in 2.4.0. I use that to figure out to flush or not. I could add a field to the Tpm2Handle structure noflush and set that. I was hoping to not put that burden on the programmer. Centos 8 is EOL IIUC, do we still need to support this with new systemd?
> > 
> > 
> > CentOS Stream 8 is not EOL so we'd still like to support it
> 
> Really? It says Dec 31 2021 here, is that wrong?
> 
>     * https://www.centos.org/centos-linux-eol/

CentOS 8 != CentOS Stream 8. C8S will be supported at least until May 2024 (https://wiki.centos.org/About/Product).

williamcroberts commented:
> > > > So Centos 8 is failing because it is on tpm2-tss  2.3.2 which predates Esys_TR_GetTPMHandle. That function was added in 2.4.0. I use that to figure out to flush or not. I could add a field to the Tpm2Handle structure noflush and set that. I was hoping to not put that burden on the programmer. Centos 8 is EOL IIUC, do we still need to support this with new systemd?
> > > 
> > > 
> > > CentOS Stream 8 is not EOL so we'd still like to support it
> > 
> > 
> > Really? It says Dec 31 2021 here, is that wrong?
> > ```
> > * https://www.centos.org/centos-linux-eol/
> > ```
> 
> CentOS 8 != CentOS Stream 8. C8S will be supported at least until May 2024 (https://wiki.centos.org/About/Product).

Should be fixed in the latest, thanks for pointing that out. I didn't realize their was that version of CentOS.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1664,18 +1901,39 @@ int tpm2_unseal(const char *device,
+                primary->flags |= TPM2_HANDLE_FLAGS_NOFLUSH;
+
+                log_debug("Found existing SRK key to use, deserializing ESYS_TR");
+                rc = sym_Esys_TR_Deserialize(
```
sorry to jump in again, but could you please marshal this instead of serializing? If the 'srk' is formatted as a serialized ESYS handle, then it's impossible to do this with an 'external' TPM. And you don't really gain anything by serializing instead of marshalling, as you are restricting the srk to only live at the specific reserved srk handle address, so you really don't need to retain that detail.

Basically, if this is merged using serialization then it won't be useful at all to my patch series, and it will be yet another optional parameter that tpm2_unseal() will have to support. I would really much prefer my approach of saving the marshalled key/template that I have queued up for a PR: https://github.com/ddstreet/systemd/commits/tpm2_save_primary_template

williamcroberts replied:
> sorry to jump in again, but could you please marshal this instead of serializing?

This is the marshalling and unmarshalling routines for an ESYS_TR handle

> If the 'srk' is formatted as a serialized ESYS handle, then it's impossible to do this with an 'external' TPM.

That's not true. ESAPI and tpm2-tools allow for conversions between raw `TPM2_HANDLE`'s (eg 0x81000001) and ESYS_TR representations.

> And you don't really gain anything by serializing instead of marshalling, as you are restricting the srk to only live at the specific reserved srk handle address, so you really don't need to retain that detail.

You're actually losing the whole point of serializing and deserializing an ESYS_TR. An ESYS_TR is both the `TPM2_HANDLE` and most importantly the **objects name, which prevents spoofing**. This is what provides the binding for the salted session where no one can spoof the object.

I don't see any point in saving the template, everything should move to a fixed persistent key at a handle which provide t he greatest flexibility in implementation. That gets you to a point where you don't even ever need to care about templates, any key without a password could be the persistent parent. You could shove in any ESYS_TR to use a non-standard location as well.

williamcroberts replied:
@ddstreet sent you an invite to your ieee.org email for a meeting, figured easier to discuss verbally than these long threads.

williamcroberts replied:
TL;DR
1. Creating an offline seal key TPM2B_PUBLIC and PRIVATE blobs still requires a stable parent key to apply duplication protections
    - The creating entity needs to know the expected parent on the actual TPM which is dependent on the Owner Hierarchy Seed. 
2. The primary parent key at some point needs to be resolved to an ESYS_TR, which can be done ahead of time as well, these are stable handles.
    - It's the TPM_HANDLE + name, since the parent needs to be stable, this can be generated ahead of time as needed.
    - They can be moved around (hence deserialize and serialize). However that exact object needs to be resident in wherever it's moved to.
3. Moving systemd to a model that prevents use of owner hierarchy directly and thus owner auth is ideal, as it supports workloads where owner auth and disk enrolling entity are different.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1557,30 +1755,67 @@ int tpm2_seal(const char *device,
-        _cleanup_free_ void *blob = NULL;
-        size_t max_size = sizeof(*private) + sizeof(*public), blob_size = 0;
+        k = ALIGN8(sizeof(*private)) + ALIGN8(sizeof(*public)); /* Some roughly sensible start value */
+        for (;;) {
```
i think you incorrectly rebased and your commit would reverse my commit that simplifies the marshalling here

williamcroberts replied:
Yeah bad rebase, i'll fix that.

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                                       sym_Tss2_RC_Decode(rc));
+
+        /* Did Not find SRK, indicate this by returning 0 */
+        if (cap_data->data.handles.count == 0) {
```
this is wrong; getcapabilities returns the handle at *or after* the handle address you requested. you can't just check for count == 0, you have to also check if the returned handle is the same as your requested handle

williamcroberts replied:
ahh yeah, I meant to add that back in good catch

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary_handle);
```
while this shouldn't be flushed, it should be TR_Close'd on cleanup

ddstreet commented on src/shared/tpm2-util.c:
```diff
@@ -1971,8 +2229,9 @@ int tpm2_extend_bytes(
                                 sym_Tss2_RC_Decode(rc));
 
         return 0;
-#else /* HAVE_OPENSSL */
```
why this  change to remove the comment and add newline?

williamcroberts replied:
probably a rebase issue, let me look again.

williamcroberts replied:
rebase glitch, messed up a few of the hunks, fixed.

poettering commented:
Catching up with the discussions here: @ddstreet what's the latest on your take on the SRK thing? Are you on board with this?

williamcroberts commented:
> Catching up with the discussions here: @ddstreet what's the latest on your take on the SRK thing? Are you on board with this?

@poettering Dan and I set up a meeting on Monday, would like me to add you to it?

poettering commented:
> @poettering Dan and I set up a meeting on Monday, would like me to add you to it?

sure

(and sorry for dropping the ball on this, too much work on my plate)

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+ * Details can be found here:
+ *   - https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
+ */
+#define SRK_HANDLE 0x81000001
```
suffix with `U`, or enclose in UINT32_C(), so that this clearly is unsigned (because otherwise it leaves in this netherworld where it is too large for a 32bit signed value, and hence could either be a 32bit unsigned value or a 64bit signed value and noone knows).

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+        ESYS_TR primary_handle = ESYS_TR_NONE;
+        _cleanup_(Esys_Freep) TPMS_CAPABILITY_DATA *cap_data = NULL;
+
+        assert(ret_primary);
```
assert(c),

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+
+        /* Did Not find SRK, indicate this by returning 0 */
+        if (cap_data->data.handles.count == 0 || cap_data->data.handles.handle[0] != SRK_HANDLE) {
+                ret_primary->esys_handle = ESYS_TR_NONE;
```
should probably use compound initialization so that the tpm2_context is initialized to NULL? i.e.

```c
*ret_primary = (Tpm2Handle) { 
        .esys_handle = ESYS_TR_NONE,
};
```

williamcroberts replied:
I never realized this was thing and it looks like its C99 supported or with glibc.

poettering replied:
hmm, i see now that memory mgmt works differently for Tpm2Hanlde than i thought...

williamcroberts replied:
> hmm, i see now that memory mgmt works differently for Tpm2Hanlde than i thought...

Do you still want this change?

williamcroberts replied:
Yeah this would blow out the context field. I really just want to update the handle.

poettering replied:
> Do you still want this change?

no, ignore. that would leak memory...

williamcroberts commented:
> > anyone with owner auth (or if owner auth is empty, which is commonly default) can create an SRK, that's what systemd currently does. the issue is persisting the SRK at its reserved handle.
> 
> > While the owner auth is typically not set for a new TPM, it should must be one of the first things changed. An empty owner auth means anyone can just nuke your keys, NV indices, etc. Depending on how NV indices are defined, it could allow owner auth to read/write them. You also want to set up passwords for the other hierarchies.
> 
> So what does setting an owner auth actually mean effectively? 

It means the entityholding ownerAuth owns the "Owner Hierarchy" on the TPM. This means things like CreatePrimary or NVDefineSpace on the Owner Hiearchy require Owner Authorization every time. Commonly some IT entity will own the TPM and the person enrolling the disk may not be in that entity. For instance, Dual booting my corporate laptop, I cannot make use of the disk encryption features, but my machine, like others following the spec, have an SRK that can be utilized.

> I mean, effectively this is just a password, but where do we get the password from?

You can prompt for it, you can actually use GetCapability to see if the password is set. Also, note that it can also be a policy. This is where it gets tricky, so moving users to an empty auth persistent key (SRK, or really any key, it doesn't matter), means that the cryptenroll process is agnostic to both hierarchy, parent passwords, parent policies, auths, etc.

> Windows stores this in the Registry? That sounds quite bad. 

Yeah it may be stored there, and yes that is quite bad.

> Or do you intend to ask the user for it? 

You can prompt.

> it's not really clear to me how this would work IRL, i.e. for a device intended for unattended boot? What am I missing? You seem quite confident that owner auth should generally be set, but if that's the case, then what#s the logic to re-acquire it on subsequent boots?

Yeah Owner Auth should be expected be set or anyone can nuke your keys and the model must work if owner auth is set or its essentially broken for many users. The whole point of the SRK is that it avoids the need for any auth values.

Currently, MakePrimary will fail if OwnerAuth is set, and their is no work around.
In my patch series, CreatePrimary and EvictControl will fail if ownerAuth is set, BUT their is a workaround. Someone with OnwerAuth can just create the SRK with tpm2-tools (or whatever), and then it just works and cryptenroll and systemd can work without needing owner auth.

Future thoughts. This sets up systemd to be really flexible. In reality the enrollment process could include the `ESYS_TR` for any parent key at any TPM address (ie non-standard SRKs).

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        ESYS_TR_NONE,
+                        &primary_handle);
```
it's a bit confusing that this converts a "handle" to a "handle"... maybe rename return field to "primary_tr"

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        ret_primary->esys_handle = primary_handle;
```
should initialize the thing fully, no, i.e.:

```c
*ret_primary = (Tpm2Handle) {
        .tpm2_context = c,
        .esys_handle = primaryHandle,
};
```

williamcroberts replied:
We could, but it's already set from the tpm2_handle_new that creates it (the new function takes the ESAPI_CONTEXT).

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -448,7 +637,15 @@ static int tpm2_make_primary(
+        log_debug("Generating %s on the TPM2 took %s.", use_srk_model ? "SRK" : "Transient Primary Key", FORMAT_TIMESPAN(now(CLOCK_MONOTONIC) - ts, USEC_PER_MSEC));
+
+        if (use_srk_model) {
+                rc = sym_Esys_EvictControl(c->esys_context, ESYS_TR_RH_OWNER, primary->esys_handle, ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE, SRK_HANDLE, &primary->esys_handle);
```
line break, as usual pleas

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -1471,7 +1672,8 @@ int tpm2_seal(const char *device,
         _cleanup_tpm2_handle_ Tpm2Handle *primary = NULL;
-        r = tpm2_make_primary(c, &primary, 0, &primary_alg);
+
+        r = tpm2_make_primary(c, 0, !!ret_srk_buf, &primary_alg, &primary);
```
for new code we prefer if function parameters that are just literal values (i.e. "0"), get prefixed with the argument name to make it easier to read what this is:

```c
r = tpm2_make_primary(
                c,
                /* alg= / 0,
                !!ret_srk_buf,
                &primary_alg,
                &primary);
```

williamcroberts replied:
ok so if I touch a hunk that's not following the current standard, update it.

poettering replied:
yeah, basically

williamcroberts commented on src/shared/tpm2-util.c:
```diff
@@ -330,55 +340,197 @@ static int tpm2_credit_random(Tpm2Context *c) {
+                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
+                                        sym_Tss2_RC_Decode(rc));
+
+        /* Make sure the alg is what we were expecting */
```
This comment is wrong and the ReadPublic call should be gated on the caller specifying if they care about the algorithm

poettering commented on src/shared/tpm2-util.h:
```diff
@@ -67,9 +73,14 @@ Tpm2Context *tpm2_context_unref(Tpm2Context *context);
 typedef struct {
         Tpm2Context *tpm2_context;
         ESYS_TR esys_handle;
+        TPM2HandleFlags flags;
```
just wondering, but why a flags field? why not just make this a bool?

williamcroberts replied:
I thought about that initially, but since everything in this is new, I figured other uses might emerge and id lay in the plumbing to avoid having to update all callers and users at a later date.

williamcroberts replied:
changed to bool

poettering commented on src/shared/tpm2-util.h:
```diff
@@ -67,9 +73,14 @@ Tpm2Context *tpm2_context_unref(Tpm2Context *context);
 #define _cleanup_tpm2_context_ _cleanup_(tpm2_context_unrefp)
 
+typedef enum TPM2HandleFlags {
+        TPM2_HANDLE_FLAGS_NOFLUSH = 1 << 0,
```
doesn't matter much, but usually it is nicer to express things "positively" instead of "negatively". hence maybe rename this "keep" or something, rather than "noflush"?

williamcroberts replied:
I can change it, I was following the dm-ioctl.h DM_NOFLUSH_FLAG.

poettering replied:
oh that's a public kernel api. they need their stuff to be extensible without breaking structure arrangements. that doesn't apply to an internal structure in our codebase.

i mean, flags are fine, and maybe we want this to be a flag one day if we have so many of them. but as long as it is just one or two i'd just make this a bool (and once it is two a bitfield, i.e. `bool x:1;`

williamcroberts replied:
> oh that's not the best model, it's a public kernel api. they aren't great with naming and they need their stuff to be extensible without breaking structure arrangements. that doesn't apply to an internal structure in our codebase.
> 

I just hate updating code bases where a bool goes to flags and theirs like 300 spots. I guess since this is internal to tpm2 it's unlikely to happen.

> i mean, flags are fine, and maybe we want this to be a flag one day if we have so many of them. but as long as it is just one or two i'd just make this a bool (and once it is two a bitfield, i.e. `bool x:1;`

Sure, bool it is.

poettering commented:
PR looks good codewise to me. if this makes sense to @ddstreet too, I am all onboard.

if installation of an srk should be optional we could certainly add a setting for it, that you can specify when enrolling. but generally i'd rather have less settings and more "good defaults" that just work for everyone.

poettering commented:
does this also close #22129?

williamcroberts commented:
> does this also close #22129?

Maybe. Yes if you're OK saying for password protected owner hierarchies you must create the SRK externally if it doesn't exist, ie via tpm2-tools , a one shot service, etc. If not, then we need to plumb owner auth prompt if required. I'd like to avoid that because it could be a policy protecting the hierarchy and not a password.

ddstreet commented:
> PR looks good codewise to me. if this makes sense to @ddstreet too, I am all onboard.
> 
> if installation of an srk should be optional we could certainly add a setting for it, that you can specify when enrolling. but generally i'd rather have less settings and more "good defaults" that just work for everyone.

Yep we had a chat, I'm all good with this PR; for my PRs I'll work with him to get them working with serialized handles. Thanks!

williamcroberts commented:
@poettering ping?

williamcroberts commented:
Rebased on current main and fixed conflicts, however I was unable to test as mkosi is failing again to build the test image, sent an email to the list.

@poettering any love on this PR yet :-p

williamcroberts commented:
> @ddstreet any way you can ping @poettering , I've been trying for weeks with no luck.

Nevermind, @poettering just emailed me. Thanks again folks.

poettering commented on src/cryptenroll/cryptenroll-tpm2.c:
```diff
@@ -142,7 +142,8 @@ int enroll_tpm2(struct crypt_device *cd,
         _cleanup_(json_variant_unrefp) JsonVariant *v = NULL, *signature_json = NULL;
         _cleanup_(erase_and_freep) char *base64_encoded = NULL;
-        size_t secret_size, blob_size, hash_size, pubkey_size = 0;
+        _cleanup_(freep) void *srk_buf = NULL;
```
doesn't matter, but free() is so fundamental, so we actually do grant it a macro of its own: `_cleanup_free_`...

but it's a weird rule, one we probably should get rid of, so maybe ignore this comment.

poettering commented on src/shared/creds-util.c:
```diff
@@ -1033,6 +1035,10 @@ int decrypt_credential_and_warn(
+                /*
+                 * TODO: Add the SRK data to the credential structure so it can be plumbed
+                 * through and used to verify the TPM session.
+                 */
```
hmm, so we have this rule that we use `/* */` for explanatory comments, and `//` for "fixme" style comments. This is one of the latter, hence should use //. but doesn't matter much

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -395,15 +551,50 @@ static int tpm2_make_primary(
+
+                        if (alg != 0 && alg != got_alg)
+                                log_warning("Caller asked for specific algorithm %u, but existing SRK is %u, ignoring",
+                                                alg, got_alg);
```
we'd usually align broken lines like this to the character after hte `(` in the preceeding line, but dosn't matter

poettering commented:
lgtm, my nits don#t really matter. let's merge. they cna be fixed later, or not, doesn#t matter

williamcroberts commented:
> lgtm, my nits don#t really matter. let's merge. they cna be fixed later, or not, doesn#t matter

I can update these nits next week, more travel, then back. Thanks @poettering and @ddstreet

arianvp commented on src/shared/creds-util.c:
```diff
@@ -703,7 +703,9 @@ int encrypt_credential_and_warn(
                               &tpm2_pcr_bank,
-                              &tpm2_primary_alg);
+                              &tpm2_primary_alg,
+                              /* ret_srk_buf= */ NULL,
```
Is there a specific reason why for systemd-creds we're not storing the sealing key under the SRK?  It would benefit from the same benefits AFAIK

williamcroberts replied:
I can't recall any specific reason, but I didn't fully study those code paths and just kept them at their current state.
